### PR TITLE
health check for rtsp camera to pravega

### PR DESCRIPTION
**Pull request by vangork**
```
Warning  Unhealthy        33s (x3 over 39s)   kubelet            Liveness probe failed: HTTP probe failed with statuscode : 500
Normal   Killing               33s                 kubelet            Container gstreamer failed liveness probe, will be restarted
Normal   Pulling              7s (x4 over 77s)      kubelet            Pulling image "devops-repo.isus.emc.com:8116/nautilus/gstreamer:luis"
Normal   Pulled               7s (x4 over 77s)      kubelet            Successfully pulled image "devops-repo.isus.emc.com:8116/nautilus/gstreamer:luis"
Normal   Created             6s (x4 over 77s)     kubelet            Created container gstreamer
Normal   Started              6s (x4 over 77s)     kubelet            Started container gstreamer
Warning  BackOff             5s (x3 over 20s)     kubelet            Back-off restarting failed container     
```

### CODE REVIEW

File: gst-plugin-pravega/src/seekable_byte_stream_writer.rs [hunk 1]
```diff
@@ -47,7 +47,7 @@ impl Write for SeekableByteWriter {
     }
 }
 
-impl Seek  for SeekableByteWriter {
+impl Seek for SeekableByteWriter {
```
Comment (claudiofahey) [inline gst-plugin-pravega/src/seekable_byte_stream_writer.rs:50]: This should be a separate PR.

File: k8s/charts/rtsp-camera-simulator/.helmignore [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# Patterns to ignore when building packages.
```
Comment (claudiofahey) [inline k8s/charts/rtsp-camera-simulator/.helmignore:1]: Create a separate PR for everything related to the RTSP camera simulator.

File: k8s/charts/rtsp-camera-simulator/values.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+image:
+  repository: "devops-repo.isus.emc.com:8116/nautilus/gstreamer:pravega-dev"
+  pullPolicy: "Always"
+
+camera:
+  user: "admin"
+  password: "password"
```
Comment (claudiofahey) [inline k8s/charts/rtsp-camera-simulator/values.yaml:7]: Add a CR to the end of the file. There are others missing CR also.

File: python_apps/rtsp-camera-to-pravega-probe.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
+#!/usr/bin/env python3
```
Comment (claudiofahey) [inline python_apps/rtsp-camera-to-pravega-probe.py:1]: Since you copied this file, I cannot easily see the changes. Rename this file back to rtsp-camera-to-pravega.py.

File: python_apps/rtsp-camera-to-pravega-probe.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+        self.value = value
+        self.update_at = int(time.time())
+    
+    def to_string(self):
+        return "value: %u at %u seconds since epoch" % (self.value, self.update_at)
+    
+    def is_uptodate(self):
+        return int(time.time()) - self.update_at < self.update_tolerance
+
+def ProbeHttpServer():
```
Comment (claudiofahey) [inline python_apps/rtsp-camera-to-pravega-probe.py:98]: Rename this to `start_http_server`.

File: python_apps/rtsp-camera-to-pravega-probe.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    hostname = '0.0.0.0'
```
Comment (claudiofahey) [inline python_apps/rtsp-camera-to-pravega-probe.py:99]: Add these as default arguments to the function like `def start_http_server(hostname='0.0.0.0', port=8080)`.

File: python_apps/rtsp-camera-to-pravega-probe.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    port = 8080
+    httpd = HTTPServer((hostname, port), ProbeHttpHandler)
+    def serve_forever(httpd):
+        with httpd:  # to make sure httpd.server_close is called
+            httpd.serve_forever()
+
+    thread = Thread(target=serve_forever, args=(httpd, ))
+    thread.setDaemon(True)
```
Comment (claudiofahey) [inline python_apps/rtsp-camera-to-pravega-probe.py:107]: Add a comment explaining why setDaemon is required.

File: python_apps/rtsp-camera-to-pravega-probe.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+    parser.add_argument("--fakesink", type=str2bool, default=False)
+    parser.add_argument("--fakesource", type=str2bool, default=False)
+    parser.add_argument("--fragment-duration-ms", type=int, default=1)
+    parser.add_argument("--keycloak-service-account-file")
+    parser.add_argument("--log-level", type=int, default=logging.INFO, help="10=DEBUG,20=INFO")
+    parser.add_argument("--pravega-controller-uri", default="tcp://127.0.0.1:9090")
+    parser.add_argument("--pravega-scope", required=True)
+    parser.add_argument("--pravega-stream", required=True)
+    parser.add_argument("--pravega-buffer-size", type=int, default=1024, help='Pravega writer buffer size in bytes')
+    parser.add_argument("--healthy-seconds", type=int, default=30)
```
Comment (claudiofahey) [inline python_apps/rtsp-camera-to-pravega-probe.py:163]: Add an option to enable/disable the HTTP server, with a default of false.

Comment (claudiofahey) [inline python_apps/rtsp-camera-to-pravega-probe.py:163]: Keep this list of arguments in alphabetical order.

**Review by claudiofahey — CHANGES_REQUESTED** [review]
This looks good. Please make a few changes.

File: gst-plugin-pravega/src/seekable_byte_stream_writer.rs (hunk 1 above)
Comment (vangork) [reply gst-plugin-pravega/src/seekable_byte_stream_writer.rs]: reverted

File: k8s/charts/rtsp-camera-simulator/.helmignore (hunk 2 above)
Comment (vangork) [reply k8s/charts/rtsp-camera-simulator/.helmignore]: deleted

File: k8s/charts/rtsp-camera-simulator/values.yaml (hunk 3 above)
Comment (vangork) [reply k8s/charts/rtsp-camera-simulator/values.yaml]: added

File: python_apps/rtsp-camera-to-pravega-probe.py (hunk 4 above)
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega-probe.py]: done

File: python_apps/rtsp-camera-to-pravega-probe.py (hunk 5 above)
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega-probe.py]: done

File: python_apps/rtsp-camera-to-pravega-probe.py (hunk 6 above)
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega-probe.py]: changed accordingly

File: python_apps/rtsp-camera-to-pravega-probe.py (hunk 7 above)
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega-probe.py]: added

File: python_apps/rtsp-camera-to-pravega-probe.py (hunk 8 above)
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega-probe.py]: add healthy-probe property

Comment (vangork) [reply python_apps/rtsp-camera-to-pravega-probe.py]: changed accordingly

**Review by claudiofahey — CHANGES_REQUESTED** [review]

File: python_apps/rtsp-camera-to-pravega.py [hunk 9]
```diff
@@ -23,12 +23,13 @@
...
 import time
 import traceback
+from threading import Thread
+from http.server import HTTPServer, BaseHTTPRequestHandler
 
 import gi
 gi.require_version("Gst", "1.0")
 from gi.repository import GObject, Gst
 
-
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py]: changed

File: python_apps/rtsp-camera-to-pravega.py [hunk 10]
```diff
@@ -50,14 +51,78 @@ def bus_call(bus, message, loop):
...
+        self.wfile.write(msg.encode('utf-8'))
+
+    # Any code greater than or equal to 200 and less than 400 indicates success. Any other code indicates failure
+    # https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/
+    def do_GET(self):
+        if self.path.lower() == "/ishealthy":
+            if ProbeHttpHandler.ts_set.is_uptodate():
+                self.send_code_msg(200, "OK")
+            else:
+                self.send_code_msg(500, "Internal Server Error")
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py]: changed

File: python_apps/rtsp-camera-to-pravega.py [hunk 11]
```diff
@@ -50,14 +51,78 @@ def bus_call(bus, message, loop):
...
 
-
 def str2bool(v):
     return bool(distutils.util.strtobool(v))
 
+def buffer_timestamp_probe(pad, info, data):
+    gst_buffer = info.get_buffer()
+    if gst_buffer:
+        data.update(gst_buffer.pts)
+        logging.info("buffer_timestamp_probe: %20s:%-8s: " % (
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py]: changed

File: python_apps/rtsp-camera-to-pravega.py [hunk 12]
```diff
@@ -50,14 +51,78 @@ def bus_call(bus, message, loop):
...
+            pad.get_parent_element().name,
+            pad.name) + data.to_string()
+        )
+        
+    else:
+        logging.info("buffer_timestamp_probe: %20s:%-8s: no buffer" % (
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py]: deleted

File: python_apps/rtsp-camera-to-pravega.py [hunk 13]
```diff
@@ -50,14 +51,78 @@ def bus_call(bus, message, loop):
...
+
+    def update(self, value):
+        self.value = value
+        self.update_at = int(time.time())
+    
+    def to_string(self):
+        return "value: %u at %u seconds since epoch" % (self.value, self.update_at)
+    
+    def is_uptodate(self):
+        return int(time.time()) - self.update_at < self.update_tolerance
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py]: changed

File: k8s/charts/rtsp-camera-to-pravega/Chart.yaml [hunk 14]
```diff
@@ -0,0 +1,23 @@
+apiVersion: v2
+name: rtsp-camera-to-pravega
+description: A Helm chart for Kubernetes
```
[the remark this answers is not in the document]
Comment (vangork) [reply k8s/charts/rtsp-camera-to-pravega/Chart.yaml]: changed

File: k8s/charts/rtsp-camera-to-pravega/Chart.yaml [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+appVersion: 1.16.0
```
[the remark this answers is not in the document]
Comment (vangork) [reply k8s/charts/rtsp-camera-to-pravega/Chart.yaml]: done

File: k8s/charts/rtsp-camera-to-pravega/values.yaml [hunk 16]
```diff
@@ -0,0 +1,13 @@
+image:
+  repository: "devops-repo.isus.emc.com:8116/nautilus/gstreamer:pravega-dev"
+  pullPolicy: "Always"
+
+camera:
+  address: "10.243.57.211"
```
[the remark this answers is not in the document]
Comment (vangork) [reply k8s/charts/rtsp-camera-to-pravega/values.yaml]: done

File: python_apps/rtsp-camera-to-pravega.py [hunk 17]
```diff
@@ -50,14 +51,78 @@ def bus_call(bus, message, loop):
...
+    httpd = HTTPServer((hostname, port), ProbeHttpHandler)
+    def serve_forever(httpd):
+        with httpd:  # to make sure httpd.server_close is called
+            httpd.serve_forever()
+
+    thread = Thread(target=serve_forever, args=(httpd, ))
+    # flag the http server thread as daemon thread so that it can be abruptly stopped at shutdown
+    thread.setDaemon(True)
+    thread.start()
+    logging.info('Probe server is listening on %s:%d' % (hostname, port))
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py]: done

**Comment by claudiofahey** [comment]
I'm not sure if you are still working on this but there are still a few unresolved comments.

File: python_apps/rtsp-camera-to-pravega.py [hunk 18]
```diff
@@ -50,14 +51,78 @@ def bus_call(bus, message, loop):
...
 
-
 def on_queue_overrun(element):
     logging.warning("Queue has overflowed and data has been lost. Try increasing buffer-size-mb.")
 
-
 def str2bool(v):
     return bool(distutils.util.strtobool(v))
 
+def buffer_timestamp_probe(pad, info, data):
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py:64]: done

File: python_apps/rtsp-camera-to-pravega.py [hunk 19]
```diff
@@ -50,14 +51,78 @@ def bus_call(bus, message, loop):
...
+        self.send_response(code)
+        self.send_header('Content-Type',
+                         'text/plain; charset=utf-8')
+        self.end_headers()
+        self.wfile.write(msg.encode('utf-8'))
+
+    # Any code greater than or equal to 200 and less than 400 indicates success. Any other code indicates failure
+    # https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/
+    def do_GET(self):
+        if self.path.lower() == "/ishealthy":
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py:119]: done

File: python_apps/rtsp-camera-to-pravega.py [hunk 20]
```diff
@@ -50,14 +51,78 @@ def bus_call(bus, message, loop):
...
+            if ProbeHttpHandler.ts_set.is_uptodate():
+                self.send_code_msg(200, "OK")
+            else:
+                self.send_code_msg(500, "Internal Server Error")
+        else:
+            self.send_code_msg(404, "Not Found")
+
+    def do_POST(self):
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py:127]: removed

File: python_apps/rtsp-camera-to-pravega.py [hunk 21]
```diff
@@ -50,14 +51,78 @@ def bus_call(bus, message, loop):
...
+        self.value = value
+        self.update_at = int(time.time())
+    
+    def to_string(self):
+        return "value: %u at %u seconds since epoch" % (self.value, self.update_at)
+    
+    def is_uptodate(self):
+        return int(time.time()) - self.update_at < self.update_tolerance
+
+class ProbeHttpHandler(BaseHTTPRequestHandler):
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py:107]: done

File: python_apps/rtsp-camera-to-pravega.py [hunk 22]
```diff
@@ -81,6 +146,8 @@ def main():
     parser.add_argument("--fakesink", type=str2bool, default=False)
     parser.add_argument("--fakesource", type=str2bool, default=False)
     parser.add_argument("--fragment-duration-ms", type=int, default=1)
+    parser.add_argument("--healthy-probe", type=str2bool, default=False)
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py:153]: done

File: python_apps/rtsp-camera-to-pravega.py [hunk 23]
```diff
@@ -81,6 +146,8 @@ def main():
...
+    parser.add_argument("--healthy-seconds", type=int, default=30)
```
[the remark this answers is not in the document]
Comment (vangork) [reply python_apps/rtsp-camera-to-pravega.py:154]: done

**Review by claudiofahey — APPROVED** [review]
Looks good!
